parser-typechecker/src/Unison/Util/Bytes.hs

```diff
@@ -3,94 +3,205 @@
+
+toBase :: BE.Base -> Bytes -> Bytes
+toBase e bs = snoc empty (BE.convertToBase e (toArray bs :: ByteString) :: ByteString)
+
 toWord8s :: Bytes -> [Word8]
-toWord8s bs = catMaybes [ at i bs | i <- [0..(size bs - 1)] ]
+toWord8s bs = chunks bs >>= B.unpack
 
 fromWord8s :: [Word8] -> Bytes
-fromWord8s bs = fromByteString (B.pack bs)
+fromWord8s bs = fromArray (view $ B.pack bs :: View ByteString)
 
 instance Monoid Bytes where
   mempty = Bytes mempty
   mappend (Bytes b1) (Bytes b2) = Bytes (b1 `mappend` b2)
 
 instance Semigroup Bytes where (<>) = mappend
 
-instance T.Measured (Sum Int) B.ByteString where
+instance T.Measured (Sum Int) (View ByteString) where
   measure b = Sum (B.length b)
 
 instance Show Bytes where
   show bs = show (toWord8s bs)
 
-instance Eq Bytes where
-  b1 == b2 | size b1 == size b2 = go b1 b2
+-- Produces two lists where the chunks have the same length
+alignChunks :: B.ByteArrayAccess ba => [View ba] -> [View ba] -> ([View ba], [View ba])
+alignChunks bs1 bs2 = (cs1, cs2)
+  where
+  cs1 = alignTo bs1 bs2
+  cs2 = alignTo bs2 cs1
+  alignTo :: B.ByteArrayAccess ba => [View ba] -> [View ba] -> [View ba]
+  alignTo bs1 [] = bs1
+  alignTo []  _  = []
+  alignTo (hd1:tl1) (hd2:tl2)
+    | len1 == len2 = hd1 : alignTo tl1 tl2
+    | len1 < len2  = hd1 : alignTo tl1 (dropView len1 hd2 : tl2)
+    | otherwise    = -- len1 > len2
+                     let (hd1',hd1rem) = (takeView len2 hd1, dropView len2 hd1)
+                     in hd1' : alignTo (hd1rem : tl1) tl2
     where
-    go b1 b2 = BL.fromChunks (chunks b1) == BL.fromChunks (chunks b2)
+      len1 = B.length hd1
+      len2 = B.length hd2
+
+instance Eq Bytes where
+  b1 == b2 | size b1 == size b2 =
+    uncurry (==) (alignChunks (chunks b1) (chunks b2))
   _ == _ = False
 
 -- Lexicographical ordering
 instance Ord Bytes where
-  b1 `compare` b2 =
-    BL.fromChunks (chunks b1) `compare` BL.fromChunks (chunks b2)
+  b1 `compare` b2 = uncurry compare (alignChunks (chunks b1) (chunks b2))
+
+--
+-- Forked from: http://hackage.haskell.org/package/memory-0.15.0/docs/src/Data.ByteArray.View.html
+-- which is already one of our dependencies. Forked because the view
+-- type in the memory package doesn't expose its constructor which makes
+-- it impossible to implement take and drop.
+--
+-- Module      : Data.ByteArray.View
+-- License     : BSD-style
+-- Maintainer  : Nicolas DI PRIMA <nicolas@di-prima.fr>
+-- Stability   : stable
+-- Portability : Good
+
+view :: B.ByteArrayAccess bs => bs -> View bs
+view bs = View 0 (B.length bs) bs
+
+takeView, dropView :: B.ByteArrayAccess bs => Int -> View bs -> View bs
+takeView k (View i n bs) = View i (min k n) bs
+dropView k (View i n bs) = View (i + (k `min` n)) (n - (k `min` n)) bs
+
+data View bytes = View
+  { viewOffset :: !Int
+  , viewSize   :: !Int
+  , unView     :: !bytes
+  }
+
+instance B.ByteArrayAccess bytes => Eq (View bytes) where
+  v1 == v2 = viewSize v1 == viewSize v2 && unsafeDupablePerformIO (
```

Comment:
To be honest, this (`unsafeDupablePerformIO`) seems like the wrong way to be doing this now. GHC has primops for comparing byte arrays that don't require dangerous stuff like this.